libcore/FileUtils.vala

```diff
@@ -881,6 +881,52 @@ namespace PF.FileUtils {
         }
     }
 
+    public string get_afp_target_uri (string standard_target_uri, string uri) {
+        /* For afp:// addresses the standard target uri has the user name stripped - we need to replace it */
+        if (Uri.parse_scheme (uri) == "afp") {
+            string origin_host, origin_filename;
+            if (get_afp_user_server_and_filename (uri, out origin_filename, out origin_host)) {
+                string target_host, target_filename;
+                if (get_afp_user_server_and_filename (standard_target_uri, out target_filename, out target_host)) {
+                    return "afp://" + Path.build_path (Path.DIR_SEPARATOR_S, origin_host, target_filename);
+                }
+            }
+        } else if (Uri.parse_scheme (standard_target_uri) == "afp") {
+            /* Probably clicked on an AFP server in network:// view which gives weird gvfs uri like:
+             * network:///dnssd-domain-<host>._afpovertcp._tcp.  We have to get any missing username from
+             * the Environment */
+            string target_host, target_filename;
+            if (get_afp_user_server_and_filename (standard_target_uri, out target_filename, out target_host)) {
+                if (!target_host.contains ("@")) {
+                    /* Cannot get username from origin uri so just use Environment user name */
+                    target_host = GLib.Environment.get_user_name () + "@" + target_host.dup ();
+                    return "afp://" + Path.build_path (Path.DIR_SEPARATOR_S, target_host, target_filename);
+                }
+            }
+        }
+
+        return standard_target_uri;
+    }
+
+    private bool get_afp_user_server_and_filename (string uri, out string filename, out string user_server) {
```

Comment:
Should this throw a GLib.Error instead of returning true and false? Right now it kind of looks like we just fail silently if this doesn't work out

Reply:
I've just added some terminal output - is that enough?  This gets called every time the file updates and we do not want to repeatedly throw error dialogs.